controllers/argocd/role.go

```diff
@@ -156,6 +176,86 @@ func (r *ReconcileArgoCD) reconcileRole(name string, policyRules []v1.PolicyRule
 	return roles, nil
 }
 
+func (r *ReconcileArgoCD) reconcileRoleForSupportedNamespaces(name string, policyRules []v1.PolicyRule, cr *argoprojv1a1.ArgoCD) ([]*v1.Role, error) {
+	var roles []*v1.Role
+
+	managedClusterArgoCDNamespaces, err := r.getManagedClusterArgoCDNamespaces(cr)
+
+	// create policy rules for each supported namespace for ArgoCD Server
+	for _, sourceNamespace := range cr.Spec.Server.SourceNamespaces {
+
+		namespace := &corev1.Namespace{}
+		if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: sourceNamespace}, namespace); err != nil {
+			return nil, err
+		}
+
+		managedNamespace := false
+		// do not reconcile roles for namespaces already containing managed-by label
+		// as it already contains roles reconciled during reconcilation of ManagedNamespaces
+		for _, ns := range r.ManagedNamespaces.Items {
+			if reflect.DeepEqual(ns, namespace) {
+				managedNamespace = true
+				break
+			}
+		}
```

Comment:
Added a map of ArgoCD instance to NamespaceList under ReconcileArgocd for optimization and cleanup

Reply:
@ishitasequeira this is not exactly what i had in mind when I suggested the change, and as far as I can tell isn't really optimizing much 

(https://github.com/argoproj-labs/argocd-operator/pull/794/files#diff-2fe0a2bb5baaf6170a23a7c08adc70117adfa25b7d481923aa8626a38cefd447R45) here I think we can update `ManagedSourceNamespaces` and `ManagedNamespaces` to be of type `map[string]string` where the key is the namespace name and value can be empty. We can see [here](https://github.com/argoproj-labs/argocd-operator/blob/90eadf8356c81bf66bd1928d2354fc6e34aed3ef/controllers/argocd/argocd_controller.go#L127) that the operator repopulates the list of namespaces for a specific instance each time that instance is being reconciled. This means that it is as good as saying each instance has its own copy of `ReconcileArgoCD` (and `ManagedSourceNamespaces` and `ManagedNamespaces` by extension). So we don't need to store a list of watched source namespaces for ALL argo-cd instances in the map, just for the current instance in question.

By using a map instead of a list, instead of lines 202-207 we can do something like 
```
if _, ok := r.ManagedNamespaces[sourceNamespace]; ok {
             managedNamespace = true
    }
``` 
since map lookup is O(1) I think considering how many times the operator reconciles, overtime the savings could be huge especially if one instance is managing multiple namespaces 

The same would apply for lines 234-239 but with `r.ManagedSourceNamespaces`

Reply:
Using it the way you have described does make sense. I have addressed this in the following commit.

Also, updating `r.ManagedNamespaces` to map includes bunch of changes.. I would create a separate task for it as part of code optimization.